test/extensions/structural_identifiability.jl

```diff
@@ -0,0 +1,293 @@
+
+    # Tests using model.component style (have to make system complete first).
+    gw_osc_complt = complete(goodwind_oscillator_catalyst)
+    make_si_ode(gw_osc_complt; measured_quantities=[gw_osc_complt.M])
+    make_si_ode(gw_osc_complt; known_p=[gw_osc_complt.pₑ])
+    make_si_ode(gw_osc_complt; measured_quantities=[gw_osc_complt.M], known_p=[gw_osc_complt.pₑ])
+    make_si_ode(gw_osc_complt; measured_quantities=[gw_osc_complt.M, gw_osc_complt.E], known_p=[gw_osc_complt.pₑ])
+    make_si_ode(gw_osc_complt; measured_quantities=[gw_osc_complt.M], known_p=[gw_osc_complt.pₑ, gw_osc_complt.pₚ])
+    make_si_ode(gw_osc_complt; measured_quantities=[gw_osc_complt.M], known_p = [:pₚ])
+    make_si_ode(gw_osc_complt; measured_quantities=[gw_osc_complt.M*gw_osc_complt.E])
+end
+
+# Tests for hierarchical model with conservation laws at both top and internal levels.
+let
+    # Identifiability analysis for Catalyst model.
+    rs1 = @reaction_network rs1 begin
+        (k1, k2), X1 <--> X2
+    end
+    rs2 = @reaction_network rs2 begin
+        (k3, k4), X3 <--> X4
+    end
+    @named rs_catalyst = flatten(compose(rs1, [rs2]))
+    @unpack X1, X2, k1, k2 = rs1
+    gi_1 = assess_identifiability(rs_catalyst; measured_quantities=[X1, X2, rs2.X3], known_p=[k1])
+    li_1 = assess_local_identifiability(rs_catalyst; measured_quantities=[X1, X2, rs2.X3], known_p=[k1])
+    ifs_1 = find_identifiable_functions(rs_catalyst; measured_quantities=[X1, X2, rs2.X3], known_p=[k1])
+
+    # Identifiability analysis for Catalyst converted to StructuralIdentifiability.jl model.
+    rs_ode = make_si_ode(rs_catalyst; measured_quantities=[X1, X2, rs2.X3], known_p=[k1])
+    gi_2 = assess_identifiability(rs_ode)
+    li_2 = assess_local_identifiability(rs_ode)
+    ifs_2 = find_identifiable_functions(rs_ode)
+
+    # Identifiability analysis for StructuralIdentifiability.jl model (declare this overwrites e.g. X2 variable etc.).
+    rs_si = @ODEmodel(
+        X1'(t) = -k1*X1(t) + k2*X2(t), 
+        X2'(t) = k1*X1(t) - k2*X2(t),
+        rs2₊X3'(t) = -rs2₊k3*rs2₊X3(t) + rs2₊k4*rs2₊X4(t), 
+        rs2₊X4'(t) = rs2₊k3*rs2₊X3(t) - rs2₊k4*rs2₊X4(t),
+        y1(t) = X1,
+        y2(t) = X2,
+        y3(t) = rs2₊X3,
+        y4(t) = k1
+    )
+    gi_3 = assess_identifiability(rs_si)
+    li_3 = assess_local_identifiability(rs_si)
+    ifs_3 = find_identifiable_functions(rs_si)
+        
+    # Check outputs.
+    @test sym_dict(gi_1) == sym_dict(gi_3)
+    @test sym_dict(li_1) == sym_dict(li_3)
+    @test length(ifs_1)-2 == length(ifs_2)-2 == length(ifs_3) # In the first case, the conservation law parameter is also identifiable.
+
+    # Checks output for the SI converted version of the catalyst model.
+    # For nested systems with conservation laws, conserved quantities like Γ[1], cannot be replaced back.
+    # Hence, here you display identifiability for `Γ[1]` instead of X2.
```

Comment:
This is why I wanted to add metadata of conservation law parameters. Then I can parse the output expression, detect conservation law parameters, and replace them with their expression. But until we add that, I don't see a natural way to do it.

(in the other cases it works because the conservation laws always occur in the context of the replace parameters, e.g. `Γ[1]-X1`. I have a record of this expression that maps it to `X2`. But in some cases, e.g. when identifiable functions are found, you might just get `Γ[1]` as an identifiable function.

Reply:
Can't use make a list of the conservation law constants, and then use `get_variables` to compare against the list?

Reply:
I think there might be some way to hack it, but I'd rather just go down the metadata route, I think it would be more reliable, and also easier to implement.